dask/dataframe/tests/test_groupby.py

```diff
@@ -2520,3 +2528,91 @@ def test_groupby_with_pd_grouper():
         ddf.groupby(pd.Grouper(key="key1"))
     with pytest.raises(NotImplementedError):
         ddf.groupby(["key1", pd.Grouper(key="key2")])
+
+
+def test_empty_partitions_with_tail():
+    df = pd.DataFrame(
+        data=[
+            ["a1", "b1"],
+            ["a1", None],
+            ["a1", "b1"],
+            [None, None],
+            [None, None],
+            [None, None],
+            ["a3", "b3"],
+            ["a3", "b3"],
+            ["a5", "b5"],
+        ],
+        columns=["A", "B"],
+    )
+
+    expected = df.groupby("A")["B"].tail(1)
+    ddf = dd.from_pandas(df, npartitions=3)
+    actual = ddf.groupby("A")["B"].tail(1)
+    assert_eq(expected, actual)
+
+
+def test_empty_partitions_with_head():
+    df = pd.DataFrame(
+        data=[
+            ["a1", "b1"],
+            ["a1", None],
+            ["a1", "b1"],
+            [None, None],
+            [None, None],
+            [None, None],
+            ["a3", "b3"],
+            ["a3", "b3"],
+            ["a5", "b5"],
+        ],
+        columns=["A", "B"],
+    )
+
+    expected = df.groupby("A")["B"].head(1)
+    ddf = dd.from_pandas(df, npartitions=3)
+    actual = ddf.groupby("A")["B"].head(1)
+    assert_eq(expected, actual)
+
+
+def test_groupby_with_head():
+    df = pd.DataFrame(
+        data=[
+            ["a0", "b1"],
+            ["a0", "b2"],
+            ["a1", "b1"],
+            ["a3", "b3"],
+            ["a3", "b3"],
+            ["a5", "b5"],
+            ["a1", "b1"],
+            ["a1", "b1"],
+            ["a1", "b1"],
+        ],
+        columns=["A", "B"],
+    )
+
+    expected = df.groupby("A")["B"].head()
+    ddf = dd.from_pandas(df, npartitions=3)
+    actual = ddf.groupby("A")["B"].head()
+    assert_eq(expected, actual)
+
+
+def test_groupby_with_tail():
+    df = pd.DataFrame(
+        data=[
+            ["a0", "b1"],
+            ["a0", "b2"],
+            ["a1", "b1"],
+            ["a3", "b3"],
+            ["a3", "b3"],
+            ["a5", "b5"],
+            ["a1", "b1"],
+            ["a1", "b1"],
+            ["a1", "b1"],
+        ],
+        columns=["A", "B"],
+    )
+
+    expected = df.groupby("A")["B"].tail()
+    ddf = dd.from_pandas(df, npartitions=3)
+    actual = ddf.groupby("A")["B"].tail()
+    assert_eq(expected, actual)
```

Comment:
Are there ways to combine any of these tests?  It might be nice to avoid a little bit of duplication here.  That is subjective though, feel free to ignore this comment.

Reply:
I assume you deduplicated, by parametrizing :ok_hand: - but it looks like you forgot to delete one of these tests. You have 2 tests called `test_groupby_with_row_operations`